### ISSUE TITLE
Connection established via HP dial-back are considered outgoing

### ISSUE THREAD

**Issue by Menduist** [status-im/nim-libp2p#940]
In this case:
https://github.com/status-im/nim-libp2p/blob/56c23a286a2aea64dd89fa750ffb4381d811908e/libp2p/services/hpservice.nim#L38-L40

the created connection will be considered outgoing, even though it is actually an incoming connection

This means that a peer could fool us to make us believe we initiated the connection, even though he did

**Comment by diegomrsantos** [comment]
> This means that a peer could fool us to make us believe we initiated the connection
Could it cause any practical issues? It seems a bit strange that a peer is initiating a connection, in practical terms, but this connection should actually be incoming.

**Comment by Menduist** [comment]
Some protocols give more "trust" to outgoing peers than incoming peers, since we made the decision to dial them

**Comment by diegomrsantos** [comment]
I believe this has been fixed by https://github.com/status-im/nim-libp2p/pull/994. @Menduist ?

**Comment by Menduist** [comment]
I don't think so, the direct connection is an outgoing connection. Not sure what is the best way to fix this specific issue

**Comment by diegomrsantos** [comment]
That's true. The problem existed for the direct and the one using DCUtR, the previous PR fixed just the latter. We could use dir Inbound for this `connect`call and the other peer could check if a connection attempt is coming from a relayed peer it just dialed. I'm not sure how clean it can be done tho.

**Comment by thomaseizinger** [comment]
> That's true. The problem existed for the direct and the one using DCUtR, the previous PR fixed just the latter. We could use dir Inbound for this `connect`call and the other peer could check if a connection attempt is coming from a relayed peer it just dialed and behave like it was Outbound. I'm not sure how clean it can be done tho.

I don't know about the inner workings of `nim-libp2p` but for the hole-punch (and connection upgrade) to succeed, there needs to be some kind flag or option that is carried along within the scope of the TCP connection that all upgrades (noise, yamux, etc) following from that should actually use a different role than what the TCP socket reports. In `rust-libp2p`, we call this a `role_override`. We in fact _always_ set this flag and it is the single source of truth for what an upgrade composed with a TCP connection should do. For regular dials, it is simply also set to `Dialer`. See https://github.com/libp2p/rust-libp2p/blob/ec2258e36e20dc8e933ddc68a0f83ef3f3b48545/core/src/upgrade/apply.rs#L41-L46 in case some code helps.

There shouldn't really be any ambiguity in that regard.

Another thing to consider is, if you are planning to support QUIC, then the hole-punching works slightly different. Instead of actually dialing, we just need to fire random UDP packets at the specified port and wait for an incoming connection. The random UDP packets establish the mapping in the NAT table of the routing device which then may allow the incoming packets (depending on the specific NAT in question).

**Comment by diegomrsantos** [comment]
@thomaseizinger thanks a lot for the input. Roughly speaking, we have a `Connection` type that has a `direction` field (Inbound)/Outbound) and we are using it to determine what role the peer should assume during `Upgrade`. Most of the time the dialer has `Outbound` direction and the listener `Inbound`. In DCUtR, as both peers dialed, they both have the Outbound direction. After the underlying TCP connection has been created, but before `Upgrade`, we override its direction to Inbound for the peer with the relayed address, the B in the DCUtR spec.

However, this PR is talking about this case in the spec: 

> The protocol starts with the completion of a relay connection from A to B. Upon observing the new connection, the inbound peer (here B) checks the addresses advertised by A via identify. If that set includes public addresses, then A may be reachable by a direct connection, in which case B attempts a unilateral connection upgrade by initiating a direct connection to A.

Should we also make B assume the `Inbound` role and A the `Outbound` in unilateral connection upgrade case?
